File: qt/gen_qpushbutton.go
package qt

import (
	"unsafe"
)

type QPushButton struct {
	h          uintptr
	isSubclass bool
}

// NewQPushButton constructs a new QPushButton object.
func NewQPushButton(parent *QWidget) *QPushButton {
	ret := newQPushButton(QPushButton_new(parent.cPointer()))
	ret.isSubclass = true
	return ret
}

// NewQPushButton2 constructs a new QPushButton object.
func NewQPushButton2() *QPushButton {
	ret := newQPushButton(QPushButton_new2())
	ret.isSubclass = true
	return ret
}

// NewQPushButton3 constructs a new QPushButton object.
func NewQPushButton3(text string) *QPushButton {
	text_ms := struct_miqt_string{}
	text_ms.data = CString(text)
	text_ms.len = size_t(len(text))
	defer free(unsafe.Pointer(text_ms.data))

	ret := newQPushButton(QPushButton_new3(text_ms))
	ret.isSubclass = true
	return ret
}

// NewQPushButton4 constructs a new QPushButton object.
func NewQPushButton4(icon *QIcon, text string) *QPushButton {
	text_ms := struct_miqt_string{}
	text_ms.data = CString(text)
	text_ms.len = size_t(len(text))
	defer free(unsafe.Pointer(text_ms.data))

	ret := newQPushButton(QPushButton_new4(icon.cPointer(), text_ms))
	ret.isSubclass = true
	return ret
}

// NewQPushButton5 constructs a new QPushButton object.
func NewQPushButton5(text string, parent *QWidget) *QPushButton {
	text_ms := struct_miqt_string{}
	text_ms.data = CString(text)
	text_ms.len = size_t(len(text))
	defer free(unsafe.Pointer(text_ms.data))

	ret := newQPushButton(QPushButton_new5(text_ms, parent.cPointer()))
	ret.isSubclass = true
	return ret
}

// NewQPushButton6 constructs a new QPushButton object.
func NewQPushButton6(icon *QIcon, text string, parent *QWidget) *QPushButton {
	text_ms := struct_miqt_string{}
	text_ms.data = CString(text)
	text_ms.len = size_t(len(text))
	defer free(unsafe.Pointer(text_ms.data))

	ret := newQPushButton(QPushButton_new6(icon.cPointer(), text_ms, parent.cPointer()))
	ret.isSubclass = true
	return ret
}

func (this *QPushButton) MetaObject() *QMetaObject {
	return newQMetaObject(QPushButton_MetaObject(this.h))
}

func (this *QPushButton) Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))
	return (unsafe.Pointer)(QPushButton_Metacast(this.h, param1_Cstring))
}

func QPushButton_Tr(s string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	var _ms struct_miqt_string = QPushButton_Tr(s_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QPushButton) SizeHint() *QSize {
	_goptr := newQSize(QPushButton_SizeHint(this.h))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QPushButton) MinimumSizeHint() *QSize {
	_goptr := newQSize(QPushButton_MinimumSizeHint(this.h))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QPushButton) AutoDefault() bool {
	return (bool)(QPushButton_AutoDefault(this.h))
}

func (this *QPushButton) SetAutoDefault(autoDefault bool) {
	QPushButton_SetAutoDefault(this.h, (bool)(autoDefault))
}

func (this *QPushButton) IsDefault() bool {
	return (bool)(QPushButton_IsDefault(this.h))
}

func (this *QPushButton) SetDefault(defaultVal bool) {
	QPushButton_SetDefault(this.h, (bool)(defaultVal))
}

func (this *QPushButton) SetMenu(menu *QMenu) {
	QPushButton_SetMenu(this.h, menu.cPointer())
}

func (this *QPushButton) Menu() *QMenu {
	return newQMenu(QPushButton_Menu(this.h))
}

func (this *QPushButton) SetFlat(flat bool) {
	QPushButton_SetFlat(this.h, (bool)(flat))
}

func (this *QPushButton) IsFlat() bool {
	return (bool)(QPushButton_IsFlat(this.h))
}

func (this *QPushButton) ShowMenu() {
	QPushButton_ShowMenu(this.h)
}

func QPushButton_Tr2(s string, c string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QPushButton_Tr2(s_Cstring, c_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func QPushButton_Tr3(s string, c string, n int) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QPushButton_Tr3(s_Cstring, c_Cstring, (int)(n))
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QPushButton) callVirtualBase_MetaObject() *QMetaObject {
	return newQMetaObject(QPushButton_virtualbase_MetaObject(unsafe.Pointer(this.h)))
}

func (this *QPushButton) OnMetaObject(slot func(super func() *QMetaObject) *QMetaObject) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QPushButton_override_virtual_MetaObject(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QPushButton_MetaObject
func miqt_exec_callback_QPushButton_MetaObject(self QPushButton, cb intptr_t) *QMetaObject {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func() *QMetaObject) *QMetaObject)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	virtualReturn := gofunc((&QPushButton{h: self}).callVirtualBase_MetaObject)

	return virtualReturn.cPointer()
}

func (this *QPushButton) callVirtualBase_Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))

	return (unsafe.Pointer)(QPushButton_virtualbase_Metacast(unsafe.Pointer(this.h), param1_Cstring))
}

func (this *QPushButton) OnMetacast(slot func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QPushButton_override_virtual_Metacast(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QPushButton_Metacast
func miqt_exec_callback_QPushButton_Metacast(self QPushButton, cb intptr_t, param1 *const_char) unsafe.Pointer {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	param1_ret := param1
	slotval1 := GoString(param1_ret)

	virtualReturn := gofunc((&QPushButton{h: self}).callVirtualBase_Metacast, slotval1)

	return virtualReturn
}

File: qt/gen_qpushbutton.cpp
// +build ignore

#include <QAbstractButton>
#include <QEvent>
#include <QFocusEvent>
#include <QIcon>
#include <QKeyEvent>
#include <QMenu>
#include <QMetaObject>
#include <QMouseEvent>
#include <QObject>
#include <QPaintDevice>
#include <QPaintEvent>
#include <QPoint>
#include <QPushButton>
#include <QSize>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QStyleOptionButton>
#include <QWidget>
#include <qpushbutton.h>
#include "gen_qpushbutton.h"
class MiqtVirtualQPushButton : public virtual QPushButton {
public:
MiqtVirtualQPushButton(QWidget* parent): QPushButton(parent) {};
MiqtVirtualQPushButton(): QPushButton() {};
MiqtVirtualQPushButton(const QString& text): QPushButton(text) {};
MiqtVirtualQPushButton(const QIcon& icon, const QString& text): QPushButton(icon, text) {};
MiqtVirtualQPushButton(const QString& text, QWidget* parent): QPushButton(text, parent) {};
MiqtVirtualQPushButton(const QIcon& icon, const QString& text, QWidget* parent): QPushButton(icon, text, parent) {};

virtual ~MiqtVirtualQPushButton() = default;
// cgo.Handle value for overwritten implementation
	intptr_t handle__MetaObject = 0;
// Subclass to allow providing a Go implementation
	virtual const QMetaObject* metaObject() const override {
if (handle__MetaObject == 0) {
return QPushButton::metaObject();
}
QMetaObject* callback_return_value = miqt_exec_callback_QPushButton_MetaObject(const_cast<MiqtVirtualQPushButton*>(this), handle__MetaObject);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	QMetaObject* virtualbase_MetaObject() const {

		return (QMetaObject*) QPushButton::metaObject();

	}
// cgo.Handle value for overwritten implementation
	intptr_t handle__Metacast = 0;
// Subclass to allow providing a Go implementation
	virtual void* qt_metacast(const char* param1) override {
if (handle__Metacast == 0) {
return QPushButton::qt_metacast(param1);
}
const char* sigval1 = (const char*) param1;

		void* callback_return_value = miqt_exec_callback_QPushButton_Metacast(this, handle__Metacast, sigval1);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	void* virtualbase_Metacast(const char* param1) {

		return QPushButton::qt_metacast(param1);

	}
};
QPushButton* QPushButton_new(QWidget* parent) {
return new MiqtVirtualQPushButton(parent);
}
QPushButton* QPushButton_new2() {
return new MiqtVirtualQPushButton();
}
QPushButton* QPushButton_new3(struct miqt_string text) {
QString text_QString = QString::fromUtf8(text.data, text.len);
	return new MiqtVirtualQPushButton(text_QString);
}
QPushButton* QPushButton_new4(QIcon* icon, struct miqt_string text) {
QString text_QString = QString::fromUtf8(text.data, text.len);
	return new MiqtVirtualQPushButton(*icon, text_QString);
}
QPushButton* QPushButton_new5(struct miqt_string text, QWidget* parent) {
QString text_QString = QString::fromUtf8(text.data, text.len);
	return new MiqtVirtualQPushButton(text_QString, parent);
}
QPushButton* QPushButton_new6(QIcon* icon, struct miqt_string text, QWidget* parent) {
QString text_QString = QString::fromUtf8(text.data, text.len);
	return new MiqtVirtualQPushButton(*icon, text_QString, parent);
}
void QPushButton_virtbase(QPushButton* src
, QAbstractButton** outptr_QAbstractButton
) {
*outptr_QAbstractButton = static_cast<QAbstractButton*>(src);
}
QMetaObject* QPushButton_MetaObject(const QPushButton* self) {
	return (QMetaObject*) self->metaObject();
}
void* QPushButton_Metacast(QPushButton* self, const char* param1) {
	return self->qt_metacast(param1);
}
struct miqt_string QPushButton_Tr(const char* s) {
	QString _ret = QPushButton::tr(s);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
QSize* QPushButton_SizeHint(const QPushButton* self) {
	return new QSize(self->sizeHint());
}
QSize* QPushButton_MinimumSizeHint(const QPushButton* self) {
	return new QSize(self->minimumSizeHint());
}
bool QPushButton_AutoDefault(const QPushButton* self) {
	return self->autoDefault();
}
void QPushButton_SetAutoDefault(QPushButton* self, bool autoDefault) {
	self->setAutoDefault(autoDefault);
}
bool QPushButton_IsDefault(const QPushButton* self) {
	return self->isDefault();
}
void QPushButton_SetDefault(QPushButton* self, bool defaultVal) {
	self->setDefault(defaultVal);
}
void QPushButton_SetMenu(QPushButton* self, QMenu* menu) {
	self->setMenu(menu);
}
QMenu* QPushButton_Menu(const QPushButton* self) {
	return self->menu();
}
void QPushButton_SetFlat(QPushButton* self, bool flat) {
	self->setFlat(flat);
}
bool QPushButton_IsFlat(const QPushButton* self) {
	return self->isFlat();
}
void QPushButton_ShowMenu(QPushButton* self) {
	self->showMenu();
}
struct miqt_string QPushButton_Tr2(const char* s, const char* c) {
	QString _ret = QPushButton::tr(s, c);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
struct miqt_string QPushButton_Tr3(const char* s, const char* c, int n) {
	QString _ret = QPushButton::tr(s, c, static_cast<int>(n));
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
void QPushButton_override_virtual_MetaObject(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQPushButton*>( (QPushButton*)(self) )->handle__MetaObject = slot;
}
QMetaObject* QPushButton_virtualbase_MetaObject(const void* self) {
	return ( (const MiqtVirtualQPushButton*)(self) )->virtualbase_MetaObject();
}
void QPushButton_override_virtual_Metacast(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQPushButton*>( (QPushButton*)(self) )->handle__Metacast = slot;
}
void* QPushButton_virtualbase_Metacast(void* self, const char* param1) {
	return ( (MiqtVirtualQPushButton*)(self) )->virtualbase_Metacast(param1);
}
void QPushButton_Delete(QPushButton* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<MiqtVirtualQPushButton*>( self );
	} else {
		delete self;
	}
}
